File: src/contexts/core/kanji/domain/repositories/kanji-repository.ts
import type { Kanji } from "../entities/kanji";
import type { KanjiIdeogram } from "../valueobjects/kanji-ideogram";

abstract class KanjiRepository {

    abstract findOneByIdeogram(ideogram: KanjiIdeogram): Promise<Kanji>;

    abstract findMany(): Promise<Kanji[]>;

    abstract createOne(kanji: Kanji): Promise<void>;

    abstract updateOne(kanji: Kanji): Promise<void>;

    abstract deleteOne(kanji: KanjiIdeogram): Promise<void>;

}

export { KanjiRepository };

File: libs/kernel/src/kanjis/domain/repositories/kanji_repository.rs
use async_trait::async_trait;

use crate::{kanjis::domain::entities::{kanji::Kanji, kanji_ideogram::KanjiIdeogram}, shared::domain::errors::DomainError};


#[async_trait]
pub trait KanjiRepository: Send + Sync + 'static {
    async fn find_many(&self) -> Result<Vec<Kanji>, DomainError>;
    async fn find_by_ideogram(&self, ideogram: &KanjiIdeogram) -> Result<Kanji, DomainError>;
    async fn create_one(&self, kanji: &Kanji) -> Result<(), DomainError>;
    async fn update_one(&self, kanji: &Kanji) -> Result<(), DomainError>;
    async fn delete_one(&self, ideogram: &KanjiIdeogram) -> Result<(), DomainError>;
}